chitchat/src/state.rs

```diff
@@ -104,6 +119,122 @@ impl NodeState {
+            // We probably experienced a reset and this delta is not usable for us anymore.
+            // This is not a bug, it can happen, but we just need to ignore it!
+            info!(
+                node=?node_delta.chitchat_id,
+                from_version=node_delta.from_version_excluded,
+                last_gc_version=node_delta.last_gc_version,
+                current_last_gc_version=self.last_gc_version,
+                "received delta from the future, ignoring it"
+            );
+            return false;
+        }
+
+        if self.max_version > node_delta.last_gc_version {
+            // The GCed tombstone have all been already received.
+            // We won't miss anything by applying the delta!
+            return true;
+        }
+
+        // This delta might be missing tombstones with version within
+        // (`node_state.max_version`..`last_gc_version`].
+        //
+        // It is ok if we don't have the associated values to begin
+        // with.
+        if self.last_gc_version >= node_delta.last_gc_version {
+            return true;
+        }
+
+        if node_delta.from_version_excluded > 0 {
+            warn!(
+                node=?node_delta.chitchat_id,
+                from_version=node_delta.from_version_excluded,
+                last_gc_version=node_delta.last_gc_version,
+                current_last_gc_version=self.last_gc_version,
+                "received an inapplicable delta, ignoring it");
+        }
+
+        let Some(delta_max_version) = node_delta
+            .key_values
+            .iter()
+            .map(|key_value_mutation| key_value_mutation.version)
+            .max()
+            .or(node_delta.max_version)
+        else {
+            // This can happen if we just hit the mtu at the moment
+            // of writing the SetMaxVersion operation.
+            return false;
+        };
+
+        if delta_max_version <= self.max_version() {
+            // There is not point apply in this delta as it is not bringing us to a newer state.
+            warn!(
+                node=?node_delta.chitchat_id,
+                from_version=node_delta.from_version_excluded,
+                delta_max_version=delta_max_version,
+                last_gc_version=node_delta.last_gc_version,
+                current_last_gc_version=self.last_gc_version,
+                "received a delta that does not bring us to a fresher state, ignoring it");
+            return false;
+        }
+
+        // We are out of sync. This delta is an invitation to `reset` our state.
+        info!(
+            node=?node_delta.chitchat_id,
+            last_gc_version=node_delta.last_gc_version,
+            current_last_gc_version=self.last_gc_version,
+            "resetting node");
+        *self = NodeState::new(node_delta.chitchat_id.clone(), self.listeners.clone());
+        if let Some(max_version) = node_delta.max_version {
+            self.max_version = max_version;
```

Comment:
This relies on the fact that node_delta.max_version is only set when when node_delta does not contain any kv mutations (otherwise, apply_delta() would skip the kv mutations since they would be <= to the max_version we just set).
Maybe comment this somewhere, e.g. in NodeDelta?
It may be a bit brittle, since setting NodeDelta.max_version even though it's not necessary would break this, but hey.

Reply:
Yeah this is way too brittle. I'll change this to make it into a max.